### ISSUE TITLE
Typescript error: StrictResponse of generic is not assignable to AsyncResponseResolverReturnType

### ISSUE THREAD

**Issue by gustavhagland** [kettanaito/msw#1823]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Reproduction repository

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgCRjMAlApgZ0gOxywBo4ARLAMwEMBXAGxgCEIATATwBV2wS4ALNGDgBfOJSgQQcAOQgcAdxkBuAFDrKtfAGMYwCPjgQA1gHEsMZNXyt6WKAB5O2PAaJwsADxhYbOOAByBgEM9NQARnYOFDQMzGxcPFgAfMkAFL6sAAoQwPgwAFxwODBQeQDmAJSIqnBwUBa0UIZpDa6EWEXOuARE1QC8yTV1dYLoAHTlFhk2OXkwpGkDQ6joLr1Y4wBWOAatPW5YlZW1onCqIuoA9FdwAKKe1OB2cLQ41FOq2m7wXk9gdgAshBtMYrDY7FA4P0jGYLODbPYHAhqGgyuFaD4ivhaCBwvYROkAERXP7PLCkmyQeZEk5fH4CayIqD+GEAbVOZIBWGBoIRkLSKLRwAxWLgAEYRCcALpAA

### Reproduction steps

I'm trying to create wrappers around the http.get method to be able to couple endpoints in our backend with their return types. Here is an example of one such wrapper:

```typescript
import { HttpResponse, DefaultBodyType, http } from 'msw';

function okGetHandler<TResponse extends NonNullable<DefaultBodyType>>(endPoint: string) {
  return (response: TResponse) => {
    http.get(endPoint, () => HttpResponse.json(response))
  } 
}
```

### Current behavior

This causes a Typescript error: 
Type 'StrictResponse<TResponse>' is not assignable to type 'AsyncResponseResolverReturnType<TResponse>'.
  Type 'StrictResponse<TResponse>' is not assignable to type 'TResponse extends undefined ? Response : StrictResponse<TResponse>'

### Expected behavior

I expect StrictResponse<TResponse> to be assignable to AsyncResponseResolverReturnType<TResponse>.

There is a easy workaround which is to simply use the native Response like so:

```typescript
export function okGetHandler<TResponse extends NonNullable<DefaultBodyType>>(endPoint: string) {
    return (response: TResponse) =>
        http.get(
            url + endPoint,
            () => new Response(JSON.stringify(response), { headers: { 'Content-Type': 'application/json' } }),
        )
}
```
 this is not as readable however.

Is there some extends clause that I'm missing on TResponse?

**Comment by kettanaito** [comment]
Thanks for reporting this, @gustavhagland! I looked into this and discovered that this was indeed a bug in MSW. Opened a fix in #1824. Your example compiles without errors with that fix. Also take a look at how I put your example into a regression test and see if you'd get inspired from it in any way: https://github.com/mswjs/msw/pull/1824/files#diff-75a5caa27c79c5fbb266f10a9c4348c50f06e36ba284099859c04688ec183ddd

**Comment by gustavhagland** [comment]
Thank you for the quick response! That looks great!

**Comment by kettanaito** [comment]
## Released: v2.0.3 🎉

This has been released in v2.0.3!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.0.3)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.0.3)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
